pydatastructs/miscellaneous_data_structures/segment_tree.py

```diff
@@ -0,0 +1,146 @@
+        obj._func = func
+        obj._array = array
+        obj._root = None
+        obj._backend = backend
+        return obj
+
+    @classmethod
+    def methods(self):
+        return ['__new__', 'build', 'update',
+                'query']
+
+    @property
+    def is_ready(self):
+        return self._root is not None
+
+    def build(self):
+        recursion_stack = Stack(implementation='linked_list')
+        node = TreeNode((0, len(self._array) - 1), None, backend=self._backend)
+        node.is_root = True
+        self._root = node
+        recursion_stack.push(node)
+
+        while not recursion_stack.is_empty:
+            node = recursion_stack.peek.key
+            start, end = node.key
+            if start == end:
+                node.data = self._array[start]
+                recursion_stack.pop()
+                continue
+
+            if (node.left is not None and
+                node.right is not None):
+                recursion_stack.pop()
+                node.data = self._func((node.left.data, node.right.data))
+            else:
+                mid = (start + end) // 2
+                if node.left is None:
+                    left_node = TreeNode((start, mid), None)
+                    node.left = left_node
+                    recursion_stack.push(left_node)
+                if node.right is None:
+                    right_node = TreeNode((mid + 1, end), None)
+                    node.right = right_node
+                    recursion_stack.push(right_node)
+
+    def update(self, index, value):
+        if not self.is_ready:
+            raise ValueError("{} tree is not built yet. ".format(self) +
+                             "Call .build method to prepare the segment tree.")
+
+        recursion_stack = Stack(implementation='linked_list')
+        recursion_stack.push((self._root, None))
+
+        while not recursion_stack.is_empty:
+            node, child = recursion_stack.peek.key
+            start, end = node.key
+            if start == end:
+                self._array[index] = value
+                node.data = value
+                recursion_stack.pop()
+                if not recursion_stack.is_empty:
+                    parent_node = recursion_stack.pop()
+                    recursion_stack.push((parent_node.key[0], node))
+                continue
+
+            if child is not None:
+                node.data = self._func((node.left.data, node.right.data))
+                recursion_stack.pop()
+                if not recursion_stack.is_empty:
+                    parent_node = recursion_stack.pop()
+                    recursion_stack.push((parent_node.key[0], node))
+            else:
+                mid = (start + end) // 2
+                if start <= index and index <= mid:
+                    recursion_stack.push((node.left, None))
+                else:
+                    recursion_stack.push((node.right, None))
+
+    def _query(self, node, start, end, l, r):
```

Comment:
`if not self.is_ready` to be added here.

Reply:
Ok, will add it..